design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,137 @@
+
+Bucket deletion in the early stages of this controller design will not be addressed beyond cleaning up API objects.  Deletion of the bucket within the object store is left as an administrator responsibility.  This is to prevent the accidental destruction of data by users.
+
+## Design
+
+### Work Flow
+
+#### Assumptions
+
+- An object store has been deployed in a Kubernetes Cluster ([e.g. Rook-Ceph Object](https://rook.github.io/docs/rook/master/ceph-object.html))
+- A `Secret` (defined below) containing the Access Key and Secret Key has been created.
+- A `StorageClass` (defined below) has been created for the object store service.
+
+**Use Case: Provision a Bucket** 
+
+_As a Kubernetes user, I want leverage the Kubernetes API to create S3 buckets. I expect to get back the bucket connection information in a ConfigMap._
+ 
+![work flow](./ObjectBucketClaim.png)
+
+1. The Rook Operator detects a new `ObjectBucketClaim` instance.  
+    1. The operator uses `objectBucketClaim.spec.secretName` to get the S3 access keys secret.  
+    1. The operator uses `objectBucketClaim.spec.storageClassName` to get the `Service` endpoint of the object store.
+1. The operator uses the object store endpoint and access keys for an S3 "make bucket" call.
+1. The operator creates a ConfigMap in the namespace of the `ObjectBucketClaim` with relevant connection data of the bucket.
+1. An app Pod may then mount the Secret and the ConfigMap to begin accessing the bucket. 
+
+**Use Case:** Delete an Object Bucket
+
+_As a Kubernetes user, I want to delete `ObjectBucketClaim` instances and cleanup generated API resources._
+
+1. The user deletes the `ObjectBucketClaim` via `kubectl delete ...`.
+1. The `ObjectBucketClaim` is marked for deletion and left in the foreground.
+1. The respective ConfigMap is deleted.
+1. The `ObjectBucketClaim` is garbage collected.
+
+---
+
+## API Specifications
+
+#### ObjectBucketClaim CRD
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1 
+kind: CustomResourceDefinition
+metadata:
+  name: objectbucketclaims.rook.io
+spec:
+  group: rook.io
+  version: v1alpha2
+  scope: namespaced
+  names:
+      kind: ObjectBucketClaim
+      listKind: ObjectBucketClaimList
+      singular: objectbucketclaim
+      plural: objectbucketclaims
+```
+
+#### ObjectBucketClaim
+
+```yaml
+apiVersion: rook.io/v1alpha2
+kind: ObjectBucketClaim
+metadata:
+  name: my-bucket-1
+  namespace: dev-user
+  labels:
+    rook.io/bucket-provisioner:
+    rook.io/object-bucket-claim: my-bucket-1 [1]
+spec:
+  storageClassName: some-object-store
+  secretName: my-s3-key-pair
+  generateBucketName: prefix [2]
+```
+
+1. Added by the rook operator.
+1. As with `metadata.generateName`, the operator will append a hyphen followed by random characters to the string given here.
```

Comment:
> Is generateBucketName optional? Or we will always generate the bucket name?

We will always generate the name to prevent collisions in the object store.

>In general, nouns are preferred over verbs in the crd. What about calling this bucketNamePrefix?

I'm okay with that, will change.